sum type can't work with generic function type

### Describe the bug

sum type can't work with generic function type

### Expected Behavior

should work

### Current Behavior

compiler issue a error
```sh
==================
/tmp/v_1000/test_func_type_alias.13589549441546426520.tmp.c:955: error: ';' expected (got "main__Fn")
...
==================
(Use `v -cg` to print the entire error message)

builder error:
==================
C error. This should never happen.

This is a compiler bug, please report it using `v bug file.v`.

https://github.com/vlang/v/issues/new/choose

You can also use #help on Discord: https://discord.gg/vlang
```

### Reproduction Steps

test_func_type_alias.v
```v
module main

type Fn[T] = fn (string) (T)
type Fnn = fn (string) (string)

type Parser = Fnn | Fn


fn main() {
        println('hello')
}
```

### Possible Solution

N/A

### Additional Information/Context

N/A

### V version

V 0.4.0 ec35d76

### Environment details (OS name and version, etc.)

V full version: V 0.4.0 ec35d76
OS: linux, Ubuntu 22.04.2 LTS
Processor: 8 cpus, 64bit, little endian, Intel(R) Core(TM) i7-9700 CPU @ 3.00GHz

getwd: /home/mars/v/9
vexe: /HD/github/lang/v/v
vexe mtime: 2023-06-25 11:46:37

vroot: OK, value: /HD/github/lang/v
VMODULES: OK, value: /home/mars/.vmodules
VTMP: OK, value: /tmp/v_1000

Git version: git version 2.34.1
Git vroot status: 0.4-7-g44a6741b
.git/config present: true

CC version: cc (Ubuntu 11.3.0-1ubuntu1~22.04.1) 11.3.0
thirdparty/tcc status: thirdparty-linux-amd64 12f392c3

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -2893,6 +2893,11 @@ fn (mut c Checker) cast_expr(mut node ast.CastExpr) ast.Type {
 	if to_type.has_flag(.option) && from_type == ast.none_type {
 		// allow conversion from none to every option type
 	} else if to_sym.kind == .sum_type {
+		to_sym_info := to_sym.info as ast.SumType
+		if to_sym_info.generic_types.len > 0 && to_sym_info.concrete_types.len == 0 {
+			c.error('generic sumtype `${to_sym.name}` must specify type parameter, e.g. ${to_sym.name}[int]',
+				node.pos)
+		}
 		if from_type in [ast.int_literal_type, ast.float_literal_type] {
 			xx := if from_type == ast.int_literal_type { ast.int_type } else { ast.f64_type }
 			node.expr_type = c.promote_num(node.expr_type, xx)

--- a/vlib/v/checker/checker.v
+++ b/vlib/v/checker/checker.v
@@ -610,6 +610,16 @@ and use a reference to the sum type instead: `var := &${node.name}(${variant_nam
 				}
 			}
 		} else if sym.info is ast.FnType {
+			if sym.info.func.generic_names.len > 0 {
+				if !variant.typ.has_flag(.generic) {
+					c.error('generic fntype `${sym.name}` must specify generic type names, e.g. ${sym.name}[T]',
+						variant.pos)
+				}
+				if node.generic_types.len == 0 {
+					c.error('generic sumtype `${node.name}` must specify generic type names, e.g. ${node.name}[T]',
+						node.name_pos)
+				}
+			}
 			if c.table.sym(sym.info.func.return_type).name.ends_with('.${node.name}') {
 				c.error('sum type `${node.name}` cannot be defined recursively', variant.pos)
 			}

--- a/vlib/v/checker/tests/generic_sumtype_decl_err_d.out
+++ b/vlib/v/checker/tests/generic_sumtype_decl_err_d.out
@@ -1,0 +1,7 @@
+vlib/v/checker/tests/generic_sumtype_decl_err_d.vv:5:15: error: generic fntype `Fn` must specify generic type names, e.g. Fn[T]
+    3 | type Fnn = fn (string) string
+    4 |
+    5 | type Parser = Fn | Fnn
+      |               ~~
+    6 |
+    7 | fn main() {

--- a/vlib/v/checker/tests/generic_sumtype_decl_err_d.vv
+++ b/vlib/v/checker/tests/generic_sumtype_decl_err_d.vv
@@ -1,0 +1,9 @@
+type Fn[T] = fn (string) T
+
+type Fnn = fn (string) string
+
+type Parser = Fn | Fnn
+
+fn main() {
+	println('hello')
+}

--- a/vlib/v/checker/tests/generic_sumtype_decl_err_e.out
+++ b/vlib/v/checker/tests/generic_sumtype_decl_err_e.out
@@ -1,0 +1,6 @@
+vlib/v/checker/tests/generic_sumtype_decl_err_e.vv:10:10: error: generic sumtype `Parser` must specify type parameter, e.g. Parser[int]
+    8 |
+    9 | fn main() {
+   10 |     _ := Parser(f)
+      |          ~~~~~~~~~
+   11 | }

--- a/vlib/v/checker/tests/generic_sumtype_decl_err_e.vv
+++ b/vlib/v/checker/tests/generic_sumtype_decl_err_e.vv
@@ -1,0 +1,11 @@
+type Fn[T] = fn (string) T
+
+type Fnn = fn (string) string
+
+type Parser[T] = Fn[T] | Fnn
+
+fn f(x string) string { return '>> f $x' }
+
+fn main() {
+    _ := Parser(f)
+}

